### PR TITLE
Refs #35842 -- Added GitHub Actions workflow job for testing against the latest SQLite release.

### PR DESCRIPTION
**Pull request by laymonage**
#### Trac ticket number

ticket-35842

The issue described in the ticket is fixed on SQLite >= 3.47.0. Unfortunately, our current CI setup does not have a test job that uses a version of SQLite that meets the criteria. This is likely because updating the SQLite version in Python isn't very straightforward: the built-in library is included in the Python installation. To update the SQLite version, you need to compile SQLite from source and make Python use the resulting library file.

#### Branch description

This PR adds a scheduled job in our existing matrix for scheduled tests, where we also test against specific database versions (e.g. PostgreSQL). This ensures that we test against the latest release of SQLite regularly, but not necessarily on every commit/PR.

A result of the tests can be found at my fork of the repo: https://github.com/laymonage/django/actions/runs/12224553784

You can see that the `model_fields.test_jsonfield.TestQuerying.test_lookups_special_chars_double_quotes` test becomes an unexpected success. This will be fixed in a separate PR.

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [ ] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: .github/workflows/schedule_tests.yml [hunk 1]
```diff
@@ -197,6 +197,54 @@ jobs:
...
+  sqlite-latest:
+    runs-on: ubuntu-latest
+    name: SQLite latest
+    continue-on-error: true
+    steps:
+      - name: Checkout SQLite
+        uses: actions/checkout@v4
+        with:
+          repository: 'sqlite/sqlite'
+          ref: 'release'
```
[the remark this answers is not in the document]
Comment (laymonage) [reply .github/workflows/schedule_tests.yml:209]: "Arbitrary" may not be the right word, sorry if that was a bit harsh. It was not meant towards your ticket description specifically.

What I mean is that  while we "_generally_ support about 5 years after released", the final decision is still based on certain parameters as you just described: SQLite in Debian and Ubuntu, SQLite shipped in Python, and the benefits of dropping support.

Compared to other databases, these are all "moving" parameters, and the SQLite version itself is not explicitly defined in our CI setup. I'm assuming it implicitly comes from the SQLite version used in the distro that we use for the CI (happy to update https://code.djangoproject.com/wiki/CI to explicitly mention it if it's useful).

We don't really declare our "supported Debian/Ubuntu versions" (and rightly so, it's just based on the distro's lifetime, which is sensible), do we?

Correct me if I'm wrong 😄 

I'm not saying we should change that or be less generous about the support – I just think if we claim support for certain SQLite versions, it's useful to have a CI job that tests the specific SQLite version regardless of the system it's installed on (e.g. ubuntu-latest) like we do for other databases.

If that's not a good idea then I'm happy to only add the CI job for the `latest` release.

[the remark this answers is not in the document]
Comment (laymonage) [reply .github/workflows/schedule_tests.yml:209]: > Other database have strict support policies that [are mentioned in wiki](https://code.djangoproject.com/wiki/SupportedDatabaseVersions) so we can easily stick to them. This doesn't apply to SQLite so we had to create our own policy.

Yes, that's exactly the point. I'm not questioning our policy or why we create it specifically for SQLite – that makes complete sense. Since we had to create our own policy, then it should be explicitly codified – either written in the docs/wiki, or in the CI setup. Otherwise you need to dig into e.g. the tickets to find the reasoning.

> I'm not sure it's worth doing, we bump SQLite versions quite "often" in Python releases so you'd need to check and update it after each minor Python release.

If it's about updating the wiki, OK – I'm happy to leave it as it is. If it's about adding the minimum version to CI, I'm not sure why that's a problem. We do the same thing for other databases (specifying in `DatabaseFeatures. minimum_database_version` _and_ CI). With our own policy for SQLite I think it's even more crucial that we set our CI explicitly to match what we state in the docs, otherwise we're just relying on the moving targets implicitly.

[the remark this answers is not in the document]
Comment (laymonage) [reply .github/workflows/schedule_tests.yml:209]: I experimented with [running the tests on the minimum version of SQLite by building it from source](https://github.com/laymonage/django/commit/f7c181d64df9eb4b5b2f0d8cf3674d2a6b91a6c3).  Some findings...

Why we'd want this:

<details><summary>Details</summary>


- SQLite didn't have the JSON1 extension enabled by default until 3.38.0. Our minimum supported version is 3.31. This is OK, [our docs specifically say that you need to enable it](https://docs.djangoproject.com/en/5.1/ref/databases/#enabling-json1-extension-on-sqlite) (at least on SQLite < 3.38).
  -  However, [Ubuntu/Debian has enabled JSON1 in their compile rules](https://git.launchpad.net/ubuntu/+source/sqlite3/tree/debian/rules?h=applied/ubuntu/focal#n47) for a while, despite still being on 3.31 for Ubuntu Focal. 
  - As a result, our CI setup _assumes_ it is _always_ enabled, even on SQLite 3.31. Again, this is because of the implicit "parameters" of the CI i.e. the distro in this case.
  - What's the effect? Our tests are less robust. Case in point: we have the following test [that fails](https://github.com/laymonage/django/actions/runs/12254518859/job/34185607636) if you run it with a database backend that does _not_ support `JSONField`, because it's missing the `@skipUnlessDBFeature` decorator: https://github.com/django/django/blob/5e998d717f7b4220a1728bd49b66ca0162e2a6cb/tests/schema/tests.py#L2425-L2443
  - This also means we don't have coverage for all code paths where the JSON-related feature flags are not supported.
- Our supported versions policy for SQLite, as you mentioned, is based on: what's in Debian/Ubuntu (with their own patches on top), and what's specified by the official Python installers for Windows and macOS.
  - This is OK to use as the basis for making the decision on which versions to support, and I think that makes a lot of sense.
  - Yet (if possible), we _should not_ make assumptions that SQLite is always installed through those means. [Our docs only state that we support >= 3.31.0 and later](https://docs.djangoproject.com/en/5.1/ref/databases/#sqlite-notes). We don't state that you should only install Python from those sources (and it doesn't make sense to state so).
  - The assumption would not be true on many common cases, e.g. installing Python from source via pyenv, or using a different Linux distribution.
  - Our CI setup for SQLite should, at the very least, have one job that tests against the minimum supported version in an environment where it does not depend on the distribution.
  - Once we drop support for SQLite < 3.38, the above point about the JSONField feature flags may no longer be relevant, but I still think having the explicit job can be useful for the point of not implicitly relying on the distro.
    - If desired, we can also instead explicitly set [`SQLITE_OMIT_JSON `](https://www.sqlite.org/compile.html#omit_json) to simulate _not_having the extension, even on newer versions.

</details> 


Why we wouldn't want this:

<details><summary>Details</summary>


- There's a few other tests that fail in the linked CI job, all of them are because of the `SQLITE_MAX_VARIABLE_NUMBER` that defaulted to 999 in < 3.32 (and 32766 in newer versions).
  - This is also something that Debian/Ubuntu has customized even in 3.31.
  - I don't think we need to worry about this as it's not something that Django controls in the code, unlike `supports_json_field`. We can fix this by setting `-DSQLITE_MAX_VARIABLE_NUMBER=32766` in the `CFLAGS` environment variable when compiling SQLite. So this isn't really a blocker.
- I had to specifically turn on `-DSQLITE_ENABLE_DESERIALIZE=1` flag, which was not enabled by default until SQLite 3.36, otherwise I ran into an error where Python couldn't find the `sqlite3_deserialize` function.
  - This is because Python [checks for the availability of the deserialize functions at compile time](https://github.com/python/cpython/blob/cef0a90d8f3a94aa534593f39b4abf98165675b9/Modules/_sqlite/connection.c#L2176).
  - GitHub's `setup-python` action [caches the Python installation](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#hosted-tool-cache), which – if we use `ubuntu-latest`, means that Python was compiled with the preinstalled SQLite 3.37.
  - If we don't want to add the `SQLITE_ENABLE_DESERIALIZE` flag, we can work around this by using an older version of Python that didn't have support for `serialize()` ([< 3.11](https://github.com/python/cpython/commit/a7551247e7cb7010fb4735281f1afa4abeb8a9cc)), or by using `ubuntu-20.04`.
  - tldr: unless we also build Python from source, there's still _some_ dependency to the SQLite version that comes with the distro 😞
  - If someone else thinks it's worth having, I'm happy to add it. Otherwise, I'm also happy to leave it out.

</details> 

TLDR: there's no 100% true way to simulate installing the minimum SQLite version on GitHub Actions that is independent of the distro, at least not without compiling Python from source as well. I'm happy if we keep the minimum version implicit via Debian/Ubuntu.

However, I still think running the tests against the latest SQLite is worth having. The above issues are highly unlikely to happen, as they only occur if you try to have a really old version of SQLite on a newer Python/Linux distro.

By having the CI job for the latest SQLite, we'd be able to catch issues like #18900 and ticket-29182 much sooner. Otherwise we'd only know about them once the SQLite version lands in Debian/Ubuntu, which is too late.
